test/distributed/test_data_parallel.py

```diff
@@ -681,6 +681,22 @@ def test_save_replica_module(self):
         dpm = torch.nn.parallel.replicate(module, devices=[0, 1], detach=True)
         torch.save(dpm, data)
 
+    @unittest.skipIf(not TEST_CUDA, "CUDA unavailable")
+    def test_module_with_no_inputs(self):
+        # gh-37814
+        class Holder(torch.nn.Module):
+            def __init__(self):
+                super().__init__()
+                self.x = torch.nn.Parameter(torch.arange(3).float())
+
+            def forward(self):
+                return self.x
+
+        expected = torch.arange(3).float().cuda()
+        holder = dp.DataParallel(Holder().cuda())
+        output = holder()
+        for replica in output.split(3):
```

Comment:
Where does this hard-coded `3` come from?

Reply:
It matches the shape of the held parameter, I'll move it to a variable to make this clear. The split is only necessary in SPMD mode which I don't think makes sense (my comment in the PR description) but I didn't want to disable the test in SPMD mode or have it fail.

It's actually not necessary with a small modification I made in the upcoming commit.